src/daft-sql/src/functions.rs

```diff
@@ -1,635 +1,172 @@
-    /// SQL 'left()' function
-    /// # Example
-    /// ```sql
-    /// SELECT left('hello', 2);
-    /// ```
-    Left,
+/// Current feature-set
+fn check_features(func: &Function) -> SQLPlannerResult<()> {
+    if func.filter.is_some() {
+        // <agg>(..) FILTER (WHERE ..)
+        unsupported_sql_err!("Aggregation `FILTER`");
+    }
+    if func.over.is_some() {
+        // WINDOW functions
+        unsupported_sql_err!("Window functions `OVER`");
+    }
+    if !func.within_group.is_empty() {
+        // <agg>(...) WITHIN GROUP
+        unsupported_sql_err!("Aggregation `WITHIN GROUP`");
+    }
+    if func.null_treatment.is_some() {
+        // Snowflake NULL treatment
+        unsupported_sql_err!("Window function `IGNORE|RESPECT NULLS`");
+    }
+    if func.parameters != FunctionArguments::None {
+        // ClickHouse parametric functions
+        unsupported_sql_err!("Parameterized function calls");
+    }
+    Ok(())
+}
 
-    /// SQL 'right()' function
-    /// # Example
-    /// ```sql
-    /// SELECT right('hello', 2);
-    /// ```
-    Right,
-    /// SQL 'find()' function
-    /// # Example
-    /// ```sql
-    /// SELECT find('hello', 'l');
-    /// ```
-    Find,
-    /// SQL 'rpad()' function
-    /// # Example
-    /// ```sql
-    /// SELECT rpad('hello', 10, ' ');
-    /// ```
-    Rpad,
-    /// SQL 'lpad()' function
-    /// # Example
-    /// ```sql
-    /// SELECT lpad('hello', 10, ' ');
-    /// ```
-    Lpad,
-    /// SQL 'repeat()' function
-    /// # Example
-    /// ```sql
-    /// SELECT repeat('X', 2);
-    /// ```
-    Repeat,
-    // Like,
-    // Ilike,
-    /// SQL 'substring()' function
-    /// # Example
-    /// ```sql
-    /// SELECT substring('hello', 1, 2);
-    /// ```
-    Substr,
-    /// SQL 'to_date()` function
-    /// # Example
-    /// ```sql
-    /// SELECT to_date('2021-01-01', 'YYYY-MM-DD');
-    /// ```
-    ToDate,
-    /// SQL 'to_datetime()' function
-    /// # Example
-    /// ```sql
-    /// SELECT to_datetime('2021-01-01 00:00:00', 'YYYY-MM-DD HH:MM:SS');
-    /// ```
-    ToDatetime,
+/// [SQLFunction] extends [FunctionExpr] with basic input validation (arity-check).
+pub struct SQLFunction(FunctionExpr);
 
-    // ------------------------------------------------
-    // Aggregate Functions
-    // ------------------------------------------------
-    Max,
+/// TODOs
+///   - Use multimap for function variants.
+///   - Add more functions..
+pub struct SQLFunctions {
+    map: HashMap<String, SQLFunction>,
 }
 
-impl SQLPlanner {
-    pub(crate) fn plan_function(&self, func: &Function) -> SQLPlannerResult<ExprRef> {
-        if func.null_treatment.is_some() {
-            unsupported_sql_err!("null treatment");
-        }
-        if func.filter.is_some() {
-            unsupported_sql_err!("filter");
+impl SQLFunction {
+    /// Early demonstration of extending the [FunctionEvaluator] trait.
+    fn validate(&self, inputs: &[ExprRef]) -> SQLPlannerResult<ExprRef> {
```

Comment:
*nit*;

I think `validate` is a bit misleading. This function does more than validating, it's transforming the sql exprs into daft exprs.  A more descriptive name would better convey its role in the translation process between SQL and the internal DSL.